File: src/p0078.java
// https://projecteuler.net/problem=78

class p78 {

    private static final int MOD = 1000000;

    /**
     * Calculates if there is a set size under n for which the number of partitions is a multiple of 1000000.
     * @param n the maximum set size.
     * @return the set length if there is one, -1 otherwise.
     */
    private static int waysMod(int n) {
        int[] ways = new int[n];
        ways[0] = 1;
        for (int i = 1; i < n; i++) {
            for (int j = i; j < n; j++) {
                ways[j] = (ways[j] + ways[j - i]) % MOD;
            }
        }
        for (int k = n - 1; k > n / 2; k--) {
            if (ways[k] == 0) {
                return k;
            }
        }
        return -1;
    }

    public static void main (String[] args) {
        long s = System.nanoTime();
        for (int i = 1; ; i *= 2) {
            int result = waysMod(i);
            if (result >= 0) {
                System.out.println(result);
                System.out.println((System.nanoTime()-s)/1000000000.0);
                System.exit(0);
            }
        }
    }
}

File: src/p0078.py
# https://projecteuler.net/problem=78

import time
from functools import lru_cache

start = time.time()

@lru_cache(maxsize=None) # cache to avoid duplicating function calls
def pn(n): # number of partitions of a nonnegative integer: p(n) = p(n - 1) + p(n - 2) - p(n - 5) - p(n - 7) + p(n - 12) + ... 
    if n == 0: # base case: one way of writing 0
        return 1
    p = 0 # number of partitions
    d1 = 1 # first index
    d2 = 1 # second index
    i = 0 # offset
    sign = 1 # to do the alternation in the sum
    while True:
        i += d2
        if i <= n:
            p += sign * pn(n - i)
        else:
            break
        d2 += 2
        i += d1
        if i <= n:
            p += sign * pn(n - i)
        else:
            break
        d1 += 1
        sign = -sign
    return p % 1000000 # no need to keep the rest of the number

i = 1
while True:
    p = pn(i) # count number of partitions
    if p % 1000000 == 0: # if divisible by 1000000, we're done
        print(i)
        break
    i += 1

end = time.time()
print(end - start)